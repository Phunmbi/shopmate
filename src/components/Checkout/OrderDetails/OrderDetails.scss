.shipping{
  width: 100%;
  height: 67vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  &-main{
    width: 1000px;
    height: 63vh;
    padding: 25px 20px;
    display: flex;
    justify-content: space-between;
    flex-direction: column;
    align-items: flex-start;
    margin: 20px 0;
    border-radius: 8px;
    background-color: var(--white-two);
    h3, p{
      margin: 10px;
    }
    &__container{
      display: flex;
      flex-direction: column;
      align-self: flex-start;
      width: 100%;
      height: 70%;
      .shipping-main__details{
        display: flex;
        width: 100%;
        .shipping-main__cart{
          width: 70%;
          .shipping-main__heading, .shipping-main__entries, .shipping-main__totals{
            display: flex;
            width: 80%;
            justify-content: space-between;
            .shipping-main__initial{
              width: 40%;
            }
          }
        }
      }
    }
    &__buttons{
      display: flex;
      width: 100%;
      justify-content: space-between;
      button{
        width: 80px;
        height: 30px;
        color: white;
        background-color: var(--red-pink);
        border: none;
        font-weight: bold;
        border-radius: 20px;
        outline: none;
        cursor: pointer;
      }
    }
  }
}
