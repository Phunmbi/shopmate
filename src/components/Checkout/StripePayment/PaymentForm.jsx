import React, {Component} from 'react';
import {CardElement, injectStripe} from 'react-stripe-elements';
import "./PaymentForm.scss";

export class PaymentForm extends Component {
  submit = async () => {
    const { stripeCharge, stripe, orderId, history } = this.props;
    let {token} = await stripe.createToken({name: "Name"});
    token && stripeCharge({
      stripeToken: token.id,
      order_id:orderId,
      description: "Shopmate shopping order",
      amount: Math.round(parseInt(localStorage.getItem("cartTotal"))) * 100, // Round to the nearest pence then convert to pounds
      history
    });
  };

  render() {
    const { redirectPaymentPage, creatingStripeCharge } = this.props;
    return (
      <div className="paymentForm">
        <div className="paymentForm-action">
          <p>Complete your order with Stripe</p>
          <CardElement />
        </div>
        <div className="paymentForm-buttons">
          <button onClick={() => redirectPaymentPage()}>Back</button>
          { creatingStripeCharge ?
            <button className="payment-loading" onClick={() => this.submit()}>Paying...</button> :
            <button onClick={() => this.submit()}>Pay</button>
          }
          </div>
      </div>
    );
  }
}

export default injectStripe(PaymentForm);
