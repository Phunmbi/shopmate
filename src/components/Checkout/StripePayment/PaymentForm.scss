.paymentForm {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  &-action{
    display: flex;
    flex-direction: column;
    align-self: flex-start;
    width: 100%;
  }

  &-buttons {
    display: flex;
    width: 100%;
    justify-content: space-between;
    button {
      width: 80px;
      height: 30px;
      color: white;
      background-color: var(--red-pink);
      border: none;
      font-weight: bold;
      border-radius: 20px;
      outline: none;
      cursor: pointer;
    }
    .payment-loading{
      background-color: rgba(246, 47, 94, 0.18) !important;

    }
  }
}