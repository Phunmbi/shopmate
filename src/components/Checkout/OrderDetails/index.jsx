import React, {Component, Fragment} from 'react';
import Loading from '../../Loading/index';
import './OrderDetails.scss'

export class OrderDetails extends Component {
  renderEntries = (cart) => {
    return cart.map((eachEntry) => {
      return (<div key={eachEntry.item_id} className="shipping-main__entries">
        <p className="shipping-main__initial">{eachEntry.name}</p>
        <p>{eachEntry.quantity}</p>
        <p>£{eachEntry.subtotal}</p>
      </div>)
    });
  };

  render() {
    const {
      redirectPage,
      cart,
      selectedCost,
      selectedRegion,
      handleNext,
      creatingOrder
    } = this.props;
    return (
      <Fragment>
        {creatingOrder ?
          <Loading height="67vh" size="full" /> :
          (
            <div className="shipping">
              <div className="shipping-main">
                <h1>Checkout</h1>
                <div className="shipping-main__container">
                  <p>Order Summary</p>
                  <div className="shipping-main__details">
                    <div className="shipping-main__cart">
                      <div className="shipping-main__heading">
                        <h3 className="shipping-main__initial">Item</h3>
                        <h3>Quantity</h3>
                        <h3>Price</h3>
                      </div>
                      {this.renderEntries(cart)}
                      <div className="shipping-main__totals">
                        <h3>Total Price</h3>
                        <h3>£{localStorage.getItem('cartTotal')}</h3>
                      </div>
                    </div>
                    <div className="shipping-main_shipping">
                      <h3>Shipping Region</h3>
                      <p>{selectedRegion}</p>
                      <h3>Shipping Cost</h3>
                      <p>{selectedCost}</p>
                    </div>
                  </div>
                </div>
                <div className="shipping-main__buttons">
                  <button onClick={() => redirectPage()}>Back</button>
                  <button onClick={() => handleNext()}>Next</button>
                </div>
              </div>
            </div>
          )
        }
      </Fragment>
    );
  }
}

export default OrderDetails;
