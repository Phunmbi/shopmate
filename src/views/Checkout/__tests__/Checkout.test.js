import React from 'react';
import { Provider } from 'react-redux';
import {StripeProvider, Elements} from 'react-stripe-elements';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import CheckoutContainer,{ Checkout } from '../index';
import {PaymentForm} from "../../../components/Checkout/StripePayment/PaymentForm";
import OrderDetails from "../../../components/Checkout/OrderDetails";

describe('<ConnectedHomePage />', () => {
  const props = {
    stripeCharge: jest.fn(),
    createOrder: jest.fn(),
    getShippingCost: jest.fn(),
    getShippingRegions: jest.fn(),
    handleNext: jest.fn(),
    redirectPaymentPage: jest.fn(),
    stripe: {
      createToken: jest.fn().mockReturnValue({token:true})
    },
    cart: [
      {name: 'adeniyi', attributes: "Blue, S", item_id: 1, quantity: 3},
      {name: 'victor', attributes: "Purple, S", item_id: 2, quantity: 4}],
    history: {
      push: jest.fn()
    },
    productDetails: {
      product_id: 1,
      name: 'nigerian',
      subtotal: 455,
      thumbnail: 'thumbnail.png',
      image_2: 'image2.png',
      image: 'image.png'
    },
    shippingRegions: [
      {shipping_region: "Canada/Us", shipping_region_id: 2},
      {shipping_region: "Rest of the World", shipping_region_id: 1}
    ],
    creatingStripeCharge: false,
    shippingCost: [
      {shipping_id: 4, shipping_type: '28 days'}
    ]
  };

  const initialState = {
    checkout: {
      order_id: '43785ff',
      creatingOrder: false,
      creatingStripeCharge: false,
      stripeChargeResponse: {
        shipping_region: 'Rest of the world'
      },
      shippingRegions: [
        {shipping_region: "Canada/Us", shipping_region_id: 2},
        {shipping_region: "Rest of the World", shipping_region_id: 1}
      ],
      loadingRegions: false,
      loadingCost: false,
      shippingCost: [
        {shipping_id: 4, shipping_type: '28 days'}
      ]
    },
    shoppingCart: {
      cart: [{name: 'adeniyi', attributes: "Blue, S", item_id: 1, quantity: 3},
        {name: 'victor', attributes: "Purple, S", item_id: 2, quantity: 4}],
      cart_Id: 4,
      loadingShoppingCart: false
    },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  let wrapper;

  beforeEach(() => {
    localStorage.setItem("name", "named");
    wrapper = mount(<Checkout {...props} store={store}/>);
  });

  it('renders appropriately', () => {
    localStorage.setItem("name", "named");
    const wrapped = mount(
      <Provider store={store}>
        <MemoryRouter key="keyed">
          <CheckoutContainer {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapped).toMatchSnapshot();
  });

  it('should simulate shipping component loading', () => {
    wrapper.setProps({loadingRegions: true});

    expect(wrapper.find('Shipping').exists()).toEqual(true);
  });

  it('should simulate shipping component handleSelectRegion', () => {
    wrapper.find('Shipping').find('.shipping-main__dropdowns').find('select').at(0).simulate('change');

    expect(wrapper.find('Shipping').exists()).toEqual(true);
  });

  it('should simulate shipping component handleSubmit', () => {
    wrapper.find('Shipping').find('.shipping-main').find('button').simulate('click');

    expect(wrapper.find('Shipping').exists()).toEqual(true);
  });

  it('should simulate shipping component handleSubmit with selectedShippingCost', () => {
    wrapper.find('Shipping').setState({selectedShippingCost: 2});

    wrapper.find('Shipping').find('.shipping-main').find('button').simulate('click');

    expect(wrapper.find('Shipping').exists()).toEqual(true);
  });

  it('should simulate OrderDetails loading', () => {
    wrapper.setState({isOrderDetailsComplete: true, isOrderCreatedAndConfirmed: false});
    wrapper.setProps({creatingOrder: true});

    expect(wrapper.find('OrderDetails').exists()).toEqual(true);
  });

  it('should simulate OrderDetails redirect back button', () => {
    wrapper.setState({isOrderDetailsComplete: true, isOrderCreatedAndConfirmed: false});

    wrapper.find('OrderDetails').find('.shipping-main__buttons').find('button').at(0).simulate('click');

    expect(wrapper.find('OrderDetails').exists()).toEqual(false);
  });

  it('should simulate OrderDetails Next Button', () => {
    wrapper.setState({isOrderDetailsComplete: true, isOrderCreatedAndConfirmed: false});

    const inst = wrapper.instance();

    inst.handlePaymentNext = jest.fn();
    inst.handlePaymentNext();

    expect(inst.handlePaymentNext).toHaveBeenCalled();
  });

  it('should simulate handleNext Next Button', () => {
    wrapper.setState({isOrderDetailsComplete: true, isOrderCreatedAndConfirmed: false});

    const wrapped = mount (
      <StripeProvider stripe={null}>
        <Elements>
          <OrderDetails {...props} store={store}/>
        </Elements>
      </StripeProvider>
    );

    wrapped.find('.shipping-main__buttons').find('button').at(1).simulate('click');

    expect(wrapped.exists()).toEqual(true);
  });

  it('should simulate Stripe Payment', () => {
    const wrapped = shallow (
      <StripeProvider stripe={null}>
        <Elements>
          <Checkout {...props} store={store} />
        </Elements>
      </StripeProvider>
    );

    wrapped.setState({isOrderDetailsComplete: true, isOrderCreatedAndConfirmed: true});

    expect(wrapped.find('PaymentForm').exists()).toEqual(false);
  });

  it('should mount and simulate PaymentForm and simulate submit', () => {
    const wrapped = mount (
      <StripeProvider stripe={null}>
        <Elements>
          <PaymentForm {...props} store={store} />
        </Elements>
      </StripeProvider>
    );

    wrapped.find('.paymentForm-buttons').find('button').at(1).simulate('click');

    expect(wrapped.find('PaymentForm').exists()).toEqual(true);
  });

  it('should mount and simulate PaymentForm and simulate back', () => {
    const wrapped = mount (
      <StripeProvider stripe={null}>
        <Elements>
          <PaymentForm {...props} store={store} />
        </Elements>
      </StripeProvider>
    );

    wrapped.find('.paymentForm-buttons').find('button').at(0).simulate('click');

    expect(wrapped.find('PaymentForm').exists()).toEqual(true);
  });

  it('should mount and simulate PaymentForm and simulate payment loading', () => {
    const wrapped = mount (
      <StripeProvider stripe={null}>
        <Elements>
          <PaymentForm {...{...props, creatingStripeCharge: true}} store={store} />
        </Elements>
      </StripeProvider>
    );

    wrapped.setProps();
    wrapped.find('.paymentForm-buttons').find('button').at(1).simulate('click');

    expect(wrapped.find('PaymentForm').exists()).toEqual(true);
  });
});
