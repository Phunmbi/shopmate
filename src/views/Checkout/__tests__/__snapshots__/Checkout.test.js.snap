// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectedHomePage /> renders appropriately 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    key="keyed"
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "zv8ekm",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "zv8ekm",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Checkout)
        cart={
          Array [
            Object {
              "attributes": "Blue, S",
              "item_id": 1,
              "name": "adeniyi",
              "quantity": 3,
            },
            Object {
              "attributes": "Purple, S",
              "item_id": 2,
              "name": "victor",
              "quantity": 4,
            },
          ]
        }
        createOrder={[MockFunction]}
        creatingStripeCharge={false}
        getShippingCost={
          [MockFunction] {
            "calls": Array [
              Array [
                1,
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        getShippingRegions={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        handleNext={[MockFunction]}
        history={
          Object {
            "push": [MockFunction],
          }
        }
        productDetails={
          Object {
            "image": "image.png",
            "image_2": "image2.png",
            "name": "nigerian",
            "product_id": 1,
            "subtotal": 455,
            "thumbnail": "thumbnail.png",
          }
        }
        redirectPaymentPage={[MockFunction]}
        shippingCost={
          Array [
            Object {
              "shipping_id": 4,
              "shipping_type": "28 days",
            },
          ]
        }
        shippingRegions={
          Array [
            Object {
              "shipping_region": "Canada/Us",
              "shipping_region_id": 2,
            },
            Object {
              "shipping_region": "Rest of the World",
              "shipping_region_id": 1,
            },
          ]
        }
        stripe={
          Object {
            "createToken": [MockFunction],
          }
        }
        stripeCharge={[MockFunction]}
      >
        <Checkout
          cart={
            Array [
              Object {
                "attributes": "Blue, S",
                "item_id": 1,
                "name": "adeniyi",
                "quantity": 3,
              },
              Object {
                "attributes": "Purple, S",
                "item_id": 2,
                "name": "victor",
                "quantity": 4,
              },
            ]
          }
          createOrder={[Function]}
          creatingOrder={false}
          creatingStripeCharge={false}
          getShippingCost={[Function]}
          getShippingRegions={[Function]}
          handleNext={[MockFunction]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          loadingCost={false}
          loadingRegions={false}
          productDetails={
            Object {
              "image": "image.png",
              "image_2": "image2.png",
              "name": "nigerian",
              "product_id": 1,
              "subtotal": 455,
              "thumbnail": "thumbnail.png",
            }
          }
          redirectPaymentPage={[MockFunction]}
          shippingCost={
            Array [
              Object {
                "shipping_id": 4,
                "shipping_type": "28 days",
              },
            ]
          }
          shippingRegions={
            Array [
              Object {
                "shipping_region": "Canada/Us",
                "shipping_region_id": 2,
              },
              Object {
                "shipping_region": "Rest of the World",
                "shipping_region_id": 1,
              },
            ]
          }
          stripe={
            Object {
              "createToken": [MockFunction],
            }
          }
          stripeCharge={[Function]}
          stripeChargeResponse={
            Object {
              "shipping_region": "Rest of the world",
            }
          }
        >
          <div
            className="checkout"
          >
            <header>
              <NavBarCollapse
                history={
                  Object {
                    "push": [MockFunction],
                  }
                }
              >
                <div
                  className="navbar-collapse"
                >
                  <div
                    className="navbar-collapse__container"
                  >
                    <h3
                      onClick={[Function]}
                    >
                      SHOPMATE
                    </h3>
                  </div>
                </div>
              </NavBarCollapse>
            </header>
            <section>
              <Shipping
                createOrder={[Function]}
                getShippingCost={[Function]}
                getShippingRegions={[Function]}
                handleSubmitShippingDetails={[Function]}
                loadingCost={false}
                loadingRegions={false}
                shippingCost={
                  Array [
                    Object {
                      "shipping_id": 4,
                      "shipping_type": "28 days",
                    },
                  ]
                }
                shippingRegions={
                  Array [
                    Object {
                      "shipping_region": "Canada/Us",
                      "shipping_region_id": 2,
                    },
                    Object {
                      "shipping_region": "Rest of the World",
                      "shipping_region_id": 1,
                    },
                  ]
                }
              >
                <div
                  className="shipping"
                >
                  <div
                    className="shipping-main"
                  >
                    <h1>
                      Checkout
                    </h1>
                    <div
                      className="shipping-main__actions"
                    >
                      <p>
                        Select shipping region
                      </p>
                      <div
                        className="shipping-main__dropdowns"
                      >
                        <select
                          name="shipping-region"
                          onChange={[Function]}
                          value={1}
                        >
                          <option
                            key="2"
                            label="Canada/Us"
                            value={2}
                          >
                            Canada/Us
                          </option>
                          <option
                            key="1"
                            label="Rest of the World"
                            value={1}
                          >
                            Rest of the World
                          </option>
                        </select>
                        <select
                          name="shipping-time"
                          onChange={[Function]}
                          value=""
                        >
                          <option>
                             
                          </option>
                          <option
                            key="4"
                            label="28 days"
                            value={4}
                          >
                            28 days
                          </option>
                        </select>
                      </div>
                    </div>
                    <button
                      onClick={[Function]}
                    >
                      Next
                    </button>
                  </div>
                </div>
              </Shipping>
            </section>
            <footer
              className="footer"
            >
              <Footer>
                <div
                  className="footer-main"
                >
                  <div
                    className="footer-container"
                  >
                    <div
                      className="footer-container__categories"
                    >
                      <h3>
                        Women
                      </h3>
                      <h3>
                        Men
                      </h3>
                      <h3>
                        Kids
                      </h3>
                      <h3>
                        Shoes
                      </h3>
                      <h3>
                        Brands
                      </h3>
                    </div>
                    <div
                      className="footer-container__social"
                    >
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="instagram"
                          src="instagramIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="pinterest"
                          src="pinterestIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="twitter"
                          src="twitterIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="facebook"
                          src="facebookIcon.svg"
                        />
                      </div>
                    </div>
                    <div
                      className="footer-container__information"
                    >
                      © shopmate Ltd • Contact • Privacy Policy
                    </div>
                  </div>
                </div>
              </Footer>
            </footer>
          </div>
        </Checkout>
      </Connect(Checkout)>
    </Router>
  </MemoryRouter>
</Provider>
`;
